<p align="center">
    <img src="polecenie.png"  width="30%">
</p>

# Import niezbędnych bibliotek

In [103]:
import pandas as pd
import numpy as np
import plotly.express as px
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

# Wczytanie i przygotwanie danych oraz eksploracyjna analiza danych

1. Wczytanie pliku z nazwami kolumn

In [104]:
cols = []
with open('column_names.txt') as f:
    cols = f.readlines()
    cols = [x.split(':')[0].strip() for x in cols]

print('Liczba wczytanych nazw kolumn: ', len(cols))
print(cols)


Liczba wczytanych nazw kolumn:  1559
['height', 'width', 'aratio', 'local', 'url*images+buttons', 'url*likesbooks.com', 'url*www.slake.com', 'url*hydrogeologist', 'url*oso', 'url*media', 'url*peace+images', 'url*blipverts', 'url*tkaine+kats', 'url*labyrinth', 'url*advertising+blipverts', 'url*images+oso', 'url*area51+corridor', 'url*ran+gifs', 'url*express-scripts.com', 'url*off', 'url*cnet', 'url*time+1998', 'url*josefina3', 'url*truluck.com', 'url*clawnext+gif', 'url*autopen.com', 'url*tvgen.com', 'url*pixs', 'url*heartland+5309', 'url*meadows+9196', 'url*blue', 'url*ad+gif', 'url*area51', 'url*www.internauts.ca', 'url*afn.org', 'url*ran.org', 'url*shareware.com', 'url*baons+images', 'url*area51+labyrinth', 'url*pics', 'url*image+navigate', 'url*pool+images', 'url*carousel.org', 'url*members.accessus.net', 'url*cmp+flowsoft010', 'url*smallb+gif', 'url*www.FlowSoft.com', 'url*kandinsky', 'url*csuhayward.edu', 'url*www.yahoo.co.uk', 'url*ibitexas.com', 'url*www.haestad.com', 'url*antiw

2. Wczytanie pliku CSV ze zbiorem danych

In [105]:
df = pd.read_csv('ad.data.csv', names=cols, header=None, na_values='?')
df.fillna(df.mean(numeric_only=True), inplace=True)

print('Czy są wartości null? ', df.isnull().values.any())
df

Czy są wartości null?  False


height       width    aratio  local  url*images+buttons  \
0     125.000000  125.000000  1.000000    1.0                   0   
1      57.000000  468.000000  8.210500    1.0                   0   
2      33.000000  230.000000  6.969600    1.0                   0   
3      60.000000  468.000000  7.800000    1.0                   0   
4      60.000000  468.000000  7.800000    1.0                   0   
...          ...         ...       ...    ...                 ...   
3274  170.000000   94.000000  0.552900    0.0                   0   
3275  101.000000  140.000000  1.386100    1.0                   0   
3276   23.000000  120.000000  5.217300    1.0                   0   
3277   64.021886  155.344828  3.911953    1.0                   0   
3278   40.000000   40.000000  1.000000    1.0                   0   

      url*likesbooks.com  url*www.slake.com  url*hydrogeologist  url*oso  \
0                      0                  0                   0        0   
1                      0                  0                   0        0   
2                      0                  0                   0        0   
3                      0                  0                   0        0   
4                      0                  0                   0        0   
...                  ...                ...                 ...      ...   
3274                   0                  0                   0        0   
3275                   0                  0                   0        0   
3276                   0                  0                   0        0   
3277                   0                  0                   0        0   
3278                   0                  0                   0        0   

      url*media  ...  caption*home  caption*my  caption*your  caption*in  \
0             0  ...             0           0             0           0   
1             0  ...             0           0             0           0   
2             0  ...             0           0             0           0   
3             0  ...             0           0             0           0   
4             0  ...             0           0             0           0   
...         ...  ...           ...         ...           ...         ...   
3274          0  ...             0           0             0           0   
3275          0  ...             0           0             0           0   
3276          0  ...             0           0             0           0   
3277          0  ...             0           0             0           0   
3278          0  ...             0           0             0           0   

      caption*bytes  caption*here  caption*click  caption*for  caption*you  \
0                 0             0              0            0            0   
1                 0             0              0            0            0   
2                 0             0              0            0            0   
3                 0             0              0            0            0   
4                 0             0              0            0            0   
...             ...           ...            ...          ...          ...   
3274              0             0              0            0            0   
3275              0             0              0            0            0   
3276              0             0              0            0            0   
3277              0             0              0            0            0   
3278              0             0              0            0            0   

      target  
0        ad.  
1        ad.  
2        ad.  
3        ad.  
4        ad.  
...      ...  
3274  nonad.  
3275  nonad.  
3276  nonad.  
3277  nonad.  
3278  nonad.  

[3279 rows x 1559 columns]

In [106]:
print('Liczba wierszy: ', df.shape[0])
print('Liczba kolumn: ', df.shape[1])

Liczba wierszy:  3279
Liczba kolumn:  1559


- Liczba elementów w zbiorze danych: **3279**
- Liczba wszystkich cech (poza ostatnią kolumną zawierającą wartości docelowe ad i nonad): **1558**


In [107]:
last_col_name = df.columns[-1]
df.groupby(last_col_name).size()

target
ad.        459
nonad.    2820
dtype: int64

In [108]:
# set size of the plot

fig = px.histogram(df, x=last_col_name, title='Liczba ogłoszeń w klasach')
fig.update_layout(width=800, height=500)
fig.show()

In [109]:
df.describe()

height        width       aratio        local  url*images+buttons  \
count  3279.000000  3279.000000  3279.000000  3279.000000         3279.000000   
mean     64.021886   155.344828     3.911953     0.767157            0.004270   
std      46.703695   110.729038     5.136153     0.421740            0.065212   
min       1.000000     1.000000     0.001500     0.000000            0.000000   
25%      32.500000    90.000000     1.279400     1.000000            0.000000   
50%      64.021886   150.000000     3.911953     1.000000            0.000000   
75%      64.021886   155.344828     3.911953     1.000000            0.000000   
max     640.000000   640.000000    60.000000     1.000000            1.000000   

       url*likesbooks.com  url*www.slake.com  url*hydrogeologist      url*oso  \
count         3279.000000        3279.000000         3279.000000  3279.000000   
mean             0.011589           0.004575            0.003355     0.003965   
std              0.107042           0.067491            0.057831     0.062850   
min              0.000000           0.000000            0.000000     0.000000   
25%              0.000000           0.000000            0.000000     0.000000   
50%              0.000000           0.000000            0.000000     0.000000   
75%              0.000000           0.000000            0.000000     0.000000   
max              1.000000           1.000000            1.000000     1.000000   

         url*media  ...   caption*of  caption*home   caption*my  caption*your  \
count  3279.000000  ...  3279.000000   3279.000000  3279.000000   3279.000000   
mean      0.011589  ...     0.006099      0.004575     0.003660      0.002440   
std       0.107042  ...     0.077872      0.067491     0.060393      0.049341   
min       0.000000  ...     0.000000      0.000000     0.000000      0.000000   
25%       0.000000  ...     0.000000      0.000000     0.000000      0.000000   
50%       0.000000  ...     0.000000      0.000000     0.000000      0.000000   
75%       0.000000  ...     0.000000      0.000000     0.000000      0.000000   
max       1.000000  ...     1.000000      1.000000     1.000000      1.000000   

        caption*in  caption*bytes  caption*here  caption*click  caption*for  \
count  3279.000000    3279.000000   3279.000000    3279.000000  3279.000000   
mean      0.003050       0.006404      0.012809       0.013419     0.009759   
std       0.055148       0.079783      0.112466       0.115077     0.098320   
min       0.000000       0.000000      0.000000       0.000000     0.000000   
25%       0.000000       0.000000      0.000000       0.000000     0.000000   
50%       0.000000       0.000000      0.000000       0.000000     0.000000   
75%       0.000000       0.000000      0.000000       0.000000     0.000000   
max       1.000000       1.000000      1.000000       1.000000     1.000000   

       caption*you  
count  3279.000000  
mean      0.001525  
std       0.039026  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       0.000000  
max       1.000000  

[8 rows x 1558 columns]

# Podział na zbiór treningowy i testowy

In [110]:
X_train, X_test, y_train, y_test = train_test_split(df.iloc[:, :-1], df.iloc[:, -1], test_size=0.2, random_state=4, stratify=df.iloc[:, -1])\

print('Liczba wierszy w zbiorze treningowym: ', X_train.shape[0])
print('Liczba wierszy w zbiorze testowym: ', X_test.shape[0])

Liczba wierszy w zbiorze treningowym:  2623
Liczba wierszy w zbiorze testowym:  656


# Usuwanie nieistotnych cech
Zbiór danych zawiera ponad 1500 cech, w związku z czym konieczne jest usunięcie części z nich (tych, o małej istotności)

In [111]:
scaler = StandardScaler()
scaler.fit(X_train)

X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

pca = PCA().fit(X_train_scaled)

fig = px.line(x=np.arange(1, len(pca.explained_variance_ratio_)+1), y=np.cumsum(pca.explained_variance_ratio_))
fig.update_layout(
    title="Wariancja wyjaśniona przez kolejne składowe",
    xaxis_title="Liczba składowych",
    yaxis_title="Wariancja wyjaśniona",
    font=dict(size=18, color="#7f7f7f"),
    width=1000,
    height=500,
)
fig.show()

pca = PCA(n_components=0.98) # 98% of variance
pca.fit(X_train_scaled)

X_train_pca = pca.transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

print('Liczba cech przed PCA: ', X_train_scaled.shape[1])
print('Liczba cech po PCA: ', X_train_pca.shape[1])


Liczba cech przed PCA:  1558
Liczba cech po PCA:  401


In [112]:
X_pca = np.concatenate((X_train_pca, X_test_pca), axis=0)
df_pca = pd.DataFrame(X_pca)
df_pca['y'] = y_test
df_pca

0         1         2         3         4         5         6  \
0    -0.754350 -0.975149 -0.511795 -0.541541  0.222503 -0.564059 -0.566025   
1    -0.862017 -1.131176 -0.512022 -0.252337 -0.387597 -0.432738 -0.190775   
2    -0.929784 -1.178363 -0.602100 -0.203032 -0.332974 -0.314331 -0.365131   
3    -0.951622 -1.250347 -0.815763 -0.652246 -0.758978 -1.411628 -1.044605   
4    -0.540277 -0.606571 -0.266414 -0.254956  0.347213 -0.062578 -0.145858   
...        ...       ...       ...       ...       ...       ...       ...   
3274 -0.310422 -1.124529 -0.303233 -0.431851 -0.216468 -0.622951  0.443251   
3275 -0.521749 -0.340545 -0.163667 -0.108614 -0.163552 -0.175622 -0.193212   
3276 -1.209277 -1.790795 -1.067532 -1.054241 -1.246752 -2.615610 -1.770888   
3277 -0.495356 -0.521343 -0.213737 -0.049661 -0.121486 -0.088981 -0.096373   
3278 -0.698405 -0.738350 -0.407641 -0.135262 -0.227594 -0.195130 -0.074852   

             7         8         9  ...       392       393       394  \
0     0.966832 -0.077254 -0.047610  ...  0.706942  0.681002 -1.001187   
1    -3.969232  2.393683 -1.342298  ... -0.274776 -0.694999  0.286296   
2    -3.811398  2.301659 -0.896092  ... -0.014538 -0.051168  0.142951   
3     3.097816 -0.800268  0.310156  ...  0.091963  0.041603 -0.061699   
4    -0.414498  0.223947 -0.096579  ... -1.000224 -0.116136  0.253103   
...        ...       ...       ...  ...       ...       ...       ...   
3274 -1.403968  0.873854 -0.960473  ...  0.366459  0.121319  1.293966   
3275 -1.097993  0.651579 -0.203040  ...  0.681148  0.834533 -0.069290   
3276  8.075199 -2.485746  1.001625  ...  0.506418  0.937831  0.466021   
3277 -0.843924  0.414397 -0.168427  ... -0.010811  0.068091  0.019735   
3278 -2.275634  1.065581  0.361454  ...  1.120281 -1.223662  2.304784   

           395       396       397       398       399       400    y  
0    -0.262927  0.342206  0.072984 -0.515972 -0.409804 -0.288330  ad.  
1     0.390206 -0.412552 -0.236543 -0.433347  0.061244 -0.017933  NaN  
2    -0.029559 -0.015717  0.065545  0.044273  0.002099  0.052001  NaN  
3     0.097125 -0.045808 -0.125168 -0.038055 -0.002164 -0.311943  ad.  
4     0.250358 -0.570876  1.820575  1.320575 -0.106473  0.245650  NaN  
...        ...       ...       ...       ...       ...       ...  ...  
3274  0.085501  0.016738  0.659196 -0.296230  0.413987  1.690362  NaN  
3275 -1.043095  0.878037  0.771412 -0.842506 -1.001758 -0.017236  NaN  
3276  0.603941  0.301212  0.289851 -0.150016 -1.509627  0.632897  NaN  
3277 -0.021625 -0.003976 -0.083095 -0.016744 -0.032209  0.018468  NaN  
3278 -0.425840 -2.936141  0.694131 -0.464202  1.611694  2.362677  NaN  

[3279 rows x 402 columns]

# Modele predykcyjne do rozpoznawania reklam internetowych

### 1. Regresja liniowa

In [113]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

lr_raw = LogisticRegression(max_iter=10000)
lr_raw.fit(X_train_scaled, y_train)

y_pred_raw = lr_raw.predict(X_test_scaled)
print('Dokładność modelu dla danych nieprzetworzonych: ', 100*np.round(accuracy_score(y_test, y_pred_raw), 4), '%')

lr_pca = LogisticRegression(max_iter=10000)
lr_pca.fit(X_train_pca, y_train)

y_pred = lr_pca.predict(X_test_pca)
print('Dokładność modelu dla danych przetworzonych PCA: ', 100*np.round(accuracy_score(y_test, y_pred), 4), '%')

Dokładność modelu dla danych nieprzetworzonych:  96.95 %
Dokładność modelu dla danych przetworzonych PCA:  96.65 %
